### ISSUE TITLE
How did you trained (further trained) GPT in a seq2seq (encoder decoder) format.

### ISSUE THREAD

**Issue by ehsan-soe** [qkaren/Counterfactual-StoryRW#3]
Hi,

I read your paper and I really enjoyed it.
I have a question regarding your training process.
Since you used the gpt architecture, I wonder how did you train it in a seq2seq format rather than language modeling? Since you have an input (source) sequence and conditioned on that you generate another (target) sequence. However, gpt is a transformer decoder. 
Did you you predict the whole source+target tokens and just compute loss for the target?

Many thanks for your explanation

**Comment by ehsan-soe** [comment]
@qkaren Did you have to pass another input to the forward method of you GPTLMHead method which is a list contains the length of source sequence,  so that you will be able to zero out the loss calculated for the tokens in source?
I mean did you zero out the lm_logits associated with source sequence so that you do not count them in loss calculation?
If you can guide me or point me to where you did that in your code that would be really nice.

**Comment by qkaren** [comment]
Thank you for the question. In our work, we put the whole source+target tokens into GPT2 as a language model (instead of a seq2seq). 

During training, we zero out the lm_logits associated with source tokens. Here is the code implementing this: [here](https://github.com/qkaren/Counterfactual-StoryRW/blob/master/rewriting_main.py#L190-L205) and [here](https://github.com/qkaren/Counterfactual-StoryRW/blob/master/rewriting_main.py#L224)

Besides, Texar also supports using GPT2 in the seq2seq model, here is an [example](https://github.com/asyml/texar/tree/master/examples/gpt-2#other-use-cases).

**Comment by ehsan-soe** [comment]
@qkaren Many thanks for your reply.
I am also curious to know if you padded your inputs? since with causal LM right padding doesn't really make sense.
Or did you concat all the data and truncate them for the desired sequence length? 
or did you batchify similar length together? if yes can you point me to the code?

Thanks in advance

**Comment by qkaren** [comment]
Texar supports automatically masking etc to handle variable-length sequences in a batch. For example, [here](https://github.com/qkaren/Counterfactual-StoryRW/blob/master/rewriting_main.py#L365), by passing in `context_sequence_length`, the right-paddings of context inputs in a batch will be automatically ignored. The same for computing sequence loss ([here](https://github.com/qkaren/Counterfactual-StoryRW/blob/master/rewriting_main.py#L193)), etc
